modules/core/src/directx.cpp

```diff
@@ -1344,6 +1345,166 @@ static void __convertFromD3D11Texture2DNV(ID3D11Texture2D* pD3D11Texture2D, Outp
+    ID3D11Device *pD3D11Device = nullptr;
+    pD3D11Texture2D->GetDevice(&pD3D11Device);
+
+    if (pD3D11Device == nullptr) {
+        CV_Error(cv::Error::StsNullPtr, "OpenCL: GetDevice returns empty pD3D11Device in convertFromD3D11Texture2DtoCLMem");
+    }
+
+    if (!device.isExtensionSupported("cl_khr_d3d11_sharing")) {
+        CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: Device doesn't support cl_khr_d3d11_sharing extension");
+    }
+
+    cl_context_properties properties[] = {
+            CL_CONTEXT_PLATFORM, (cl_context_properties)platform,
+            CL_CONTEXT_D3D11_DEVICE_KHR, (cl_context_properties)(pD3D11Device),
+            CL_CONTEXT_INTEROP_USER_SYNC, CL_FALSE,
+            NULL, NULL
+    };
+    cl_context context = NULL;
+    context = clCreateContext(properties, 1, &cl_device, NULL, NULL, &status);
+    if (status != CL_SUCCESS) {
+        CV_Error(status, "OpenCL: clCreateContext failed");
+    }
+    OpenCLExecutionContext clExecCtx;
+    try {
+        clExecCtx = OpenCLExecutionContext::create(platformName, platform, context, cl_device);
+        clExecCtx.getContext().setUserContext(std::make_shared<OpenCL_D3D11>(platform, pD3D11Device));
+    } catch (...) {
+        pD3D11Device->Release();
+        throw;
+    }
+    pD3D11Device->Release();
+    return clExecCtx;
+}
+#endif
+
+#if defined(HAVE_DIRECTX) && defined(HAVE_OPENCL)
+static void __convertFromD3D11Texture2DDev(ID3D11Texture2D* pD3D11Texture2D, Device& device, OutputArray dst) {
+    OpenCLExecutionContext clExecCtx =
+        getExecContextbyD3D11TextureAndDevice(pD3D11Texture2D, device);
+    if (clExecCtx.empty()) {
+        CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: Execution Context is empty");
+    }
+    clExecCtx.bind();
+    auto context = const_cast<Context&>(clExecCtx.getContext());
+
+    OpenCL_D3D11* impl = context.getUserContext<OpenCL_D3D11>().get();
+    if (impl == nullptr) {
+        CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: Context initialized without DirectX interoperability");
+    }
+
+    D3D11_TEXTURE2D_DESC desc = { 0 };
+    pD3D11Texture2D->GetDesc(&desc);
+
+    UMat u = dst.getUMat();
+    // TODO Add support for roi
+    CV_Assert(u.offset == 0);
+    CV_Assert(u.isContinuous());
+    int textureType = getTypeFromDXGI_FORMAT(desc.Format);
+    CV_Assert(textureType >= 0);
+
+    u.create(Size(desc.Width, desc.Height), textureType);
+    cl_mem clBuffer = (cl_mem)u.handle(ACCESS_READ);
+
+    cl_int status = 0;
+    cl_mem clImage = impl->clCreateFromD3D11Texture2DKHR((cl_context)context.ptr(), CL_MEM_READ_ONLY, pD3D11Texture2D, 0, &status);
+    if (status != CL_SUCCESS) {
+        CV_Error(status, "OpenCL: clCreateFromD3D11Texture2DKHR failed during image memory creation");
+    }
+#ifdef HAVE_DIRECTX_NV12
+    cl_mem clImageUV = 0;
+    if(DXGI_FORMAT_NV12 == desc.Format) {
+        clImageUV = impl->clCreateFromD3D11Texture2DKHR((cl_context)context.ptr(), CL_MEM_READ_ONLY, pD3D11Texture2D, 1, &status);
```

Comment:
https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clCreateFromD3D11Texture2DKHR.html

it is incorrect that you set `1` as `subresource` argument in general way
I have idea why it has worked before latest rebase:
VPL had used TextureArray before for decoding surfaces: It meant that we had only one Texture pointer which represented multiple subtextures (`subresource`s) inside it. So in this case Texture pointer was a pointer on array of textures-subresources.
And in this case when use pass parameter `0` and `1` for this `clCreateFromD3D11Texture2DKHR` then you just extract the `current` and the `next` decoding surfaces (current and the next MediaFrame implicitly) which was incorrect too.
But for some reasons it worked( may be because subresources in TextureArray initialized by zero or something other)

But after some events (Integrating to openVINO) decoding routine in VPL was changed and i applied model `Array of Textures` instead of `TexturesArray`. It means at now that instead of single Texture pointer with multiple subresources for multiple MediaFrames we have multiple Textures with 0 subresources for Multiple Frames
So  `clCreateFromD3D11Texture2DKHR` with argument `1` as `subresource` has no meaning now because the MAX number of `subresource` id in a single texture is `0` at now and represent texture by itself. So you MUST NOT  use subresource `1`  for extract UV component but should do it in another way.

**But** because, in your latter implementation you use 0 & 1: which meant single NV12 texture for current Frame and single NV12 Texture for the next Frame - `different surfaces` but not `different planes on the same surface`  then i suggest you to pass  *the same* Image twice in ocl nv12-to-grb converting function, just for experiment (lulz :) )

Reply:
Correct way is to use subuffer from `cl_mem`  which represents correct data region for U and UV separatedly:
https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_API.html#clCreateSubBuffer

so we need something like that:
```
cl_mem y = clCreateSubBuffer( image, 0, h*w/2); 
cl_mem uv = clCreateSubBuffer( image, h*2/w,  h*w/2); 

ocl::ocl_convert_nv12_to_bgr(y, uv, clBuffer, (int)u.step[0], u.cols, u.rows)) {
```